File: app/src/main/java/maderski/markparkingspot/MPSNotification.java
package maderski.markparkingspot;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.v4.content.ContextCompat;
import android.support.v4.app.NotificationCompat;

/**
 * Created by Jason on 8/9/16.
 */
public class MPSNotification {
    private static final String nTAG = Notification.class.getName();
    private static final int nID = 809;

    private Context context;

    private NotificationManager nManager;
    private NotificationCompat.Builder builder;

    private String title;
    private String message;
    private PendingIntent pendingIntent;
    private PendingIntent canGetLocationIntent;
    private String actionText;

    private int color;
    private int actionIcon;


    public MPSNotification(Context context){
        this.context = context;
        this.nManager = (NotificationManager)context
                .getSystemService(Context.NOTIFICATION_SERVICE);
        this.builder = new NotificationCompat.Builder(context);
    }

    //Create notification message
    public void createMessage(boolean isAnUpdate, boolean isAButtonPress){
        boolean enabled = MPSPreferences.CanGetNewLocation(context);

        String pinLabel = "Parking Spot " + MPSPreferences.getCurrentDate(context);
        String latitude = MPSPreferences.getLatitude(context);
        String longitude = MPSPreferences.getLongitude(context);
        String accuracy = MPSPreferences.getAccuracy(context);

        title = setTitleText(enabled, isAnUpdate) + accuracy + "ft";
        message = "at " + MPSPreferences.getCurrentTime(context) +
                " on " + MPSPreferences.getCurrentDate(context);
        actionText = setActionButtonText(enabled);

        color = ContextCompat.getColor(context, R.color.colorAccent);
        actionIcon = setActionButtonIcon(enabled);

        pendingIntent = dropPinPendingIntent(latitude, longitude, pinLabel);
        canGetLocationIntent = setCanGetLocationIntent();

        if(isAButtonPress)
            buildNoVibrateNotification();
        else
            buildNotification();
    }

    //Create pending Intent for notification
    private PendingIntent dropPinPendingIntent(String latitude, String longitude, String pinLabel){
        Intent dropPinIntent = new Intent(Intent.ACTION_VIEW,
                Uri.parse("geo:0,0?q="+ latitude+","+ longitude+"("+ pinLabel+")"));
        dropPinIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        return PendingIntent.getActivity(context, 0, dropPinIntent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    //Create CanGetNewLocation Pending Intent
    private PendingIntent setCanGetLocationIntent(){
        Intent intent = new Intent(context, GetNewLocationSetter.class);

        return PendingIntent.getActivity(context, 1, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    //Set Action Button Text
    private String setActionButtonText(boolean enabled){
        if(enabled)
            return "Get new location: ON";
        else
            return "Get new location: OFF";

    }

    private int setActionButtonIcon(boolean enabled){
        if(enabled)
            return R.drawable.ic_map_pin;
        else
            return android.R.drawable.ic_menu_close_clear_cancel;
    }

    private String setTitleText(boolean enabled, boolean isAnUpdate){
        if(enabled && !isAnUpdate)
            return "Captured! w/Accuracy: ";
        else
            return "Drop pin w/Accuracy: ";
    }

    //Build notification message
    private void buildNotification(){
        builder.setContentTitle(title)
                .setContentText(message)
                .setSmallIcon(R.drawable.ic_car)
                .setAutoCancel(true)
                .setOngoing(false)
                .setContentIntent(pendingIntent)
                .setColor(color)
                .setPriority(NotificationCompat.PRIORITY_MAX)
                .addAction(actionIcon, actionText, canGetLocationIntent)
                .setDefaults(Notification.DEFAULT_VIBRATE);
        nManager.notify(nTAG, nID, builder.build());
    }

    private void buildNoVibrateNotification(){
        builder.setContentTitle(title)
                .setContentText(message)
                .setSmallIcon(R.drawable.ic_car)
                .setAutoCancel(true)
                .setOngoing(false)
                .setContentIntent(pendingIntent)
                .setColor(color)
                .setPriority(NotificationCompat.PRIORITY_MAX)
                .addAction(actionIcon, actionText, canGetLocationIntent);
        nManager.notify(nTAG, nID, builder.build());
    }
}

File: app/src/main/java/maderski/markparkingspot/MPSNotification.kt
package maderski.markparkingspot

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.support.annotation.RequiresApi
import android.support.v4.content.ContextCompat
import android.support.v4.app.NotificationCompat

/**
 * Created by Jason on 8/9/16.
 */
class MPSNotification(private val context: Context) {

    private val notificationManager: NotificationManager = context
            .getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

    private var title: String? = null
    private var message: String? = null
    private var pendingIntent: PendingIntent? = null
    private var color: Int = 0


    //Create notification message
    fun createMessage(isAnUpdate: Boolean) {
        val enabled = MPSPreferences.canGetNewLocation(context)

        val pinLabel = "Parking Spot " + MPSPreferences.getCurrentDate(context)
        val latitude = MPSPreferences.getLatitude(context)
        val longitude = MPSPreferences.getLongitude(context)
        val accuracy = MPSPreferences.getAccuracy(context)

        title = setTitleText(enabled, isAnUpdate) + accuracy + "ft"
        message = "at " + MPSPreferences.getCurrentTime(context) +
                " on " + MPSPreferences.getCurrentDate(context)
        color = ContextCompat.getColor(context, R.color.colorAccent)
        pendingIntent = dropPinPendingIntent(latitude, longitude, pinLabel)

        buildNotification()
    }

    //Create pending Intent for notification
    private fun dropPinPendingIntent(latitude: String, longitude: String, pinLabel: String): PendingIntent {
        val dropPinIntent = Intent(Intent.ACTION_VIEW,
                Uri.parse("geo:0,0?q=$latitude,$longitude($pinLabel)"))
        dropPinIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        return PendingIntent.getActivity(context, 0, dropPinIntent, PendingIntent.FLAG_UPDATE_CURRENT)
    }

    private fun setTitleText(enabled: Boolean, isAnUpdate: Boolean): String {
        return if (enabled && !isAnUpdate)
            context.getString(R.string.captured_with_accuracy)
        else
            context.getString(R.string.drop_pin_with_accuracy)
    }

    //Build notification message
    private fun buildNotification() {
        val builder = if (Build.VERSION.SDK_INT < 26) {
            NotificationCompat.Builder(context)
        } else {
            val channel = getNotificationChannel(CHANNEL_ID, CHANNEL_NAME)
            notificationManager.createNotificationChannel(channel)
            NotificationCompat.Builder(context, CHANNEL_ID)
        }

        builder.setContentTitle(title)
                .setContentText(message)
                .setSmallIcon(R.drawable.ic_car)
                .setAutoCancel(true)
                .setOngoing(false)
                .setContentIntent(pendingIntent)
                .setDefaults(Notification.DEFAULT_VIBRATE)
                .setColor(color)
                .priority = NotificationCompat.PRIORITY_DEFAULT

        notificationManager.notify(nTAG, nID, builder.build())
    }

    @RequiresApi(Build.VERSION_CODES.O)
    private fun getNotificationChannel(channelId: String, channelName: String): NotificationChannel {
        val notificationChannel = NotificationChannel(channelId,
                channelName,
                NotificationManager.IMPORTANCE_DEFAULT)
        notificationChannel.enableVibration(true)
        notificationChannel.setShowBadge(true)
        notificationChannel.importance = NotificationManager.IMPORTANCE_DEFAULT
        return notificationChannel
    }

    companion object {
        private const val nTAG = "MPSNotification"
        private const val nID = 809
        private const val CHANNEL_NAME = "Mark Parking Spot"
        private const val CHANNEL_ID = "MPSChannelID"
    }
}
